### PR TITLE
Add auto-tuning for sparse semi-structured MM operator

### PR DESCRIPTION
**Pull request by alexsamardzic**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #123742

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Comment by alexsamardzic** [comment]
Note that the generated kernels will still sometimes fail because of alignment issues, am investigating this.

**Comment by alexsamardzic** [comment]
This PR is ready for a review now.

**Comment by alexsamardzic** [comment]
Added test validating that, on SM80 arch, at least one working CUTLASS candidate kernel is produced.

File: torch/_inductor/codegen/cuda/gemm_template.py [hunk 1]
```diff
@@ -1256,12 +1376,64 @@ def _are_inputs_layout_compatible(self, layouts: List[Layout]) -> bool:
         K = max(A_size[-1], B_size[-2])
         M = A_size[-2]
         N = B_size[-1]
-        if K != A_size[-1]:
+        if K != A_size[-1] and K != 2 * A_size[-2]:
```
Comment (kadeng) [inline torch/_inductor/codegen/cuda/gemm_template.py:1379]: Why does K have to equal twice the size of the first dimension of A on this line?

File: torch/_inductor/codegen/cuda/gemm_template.py [hunk 2]
```diff
@@ -1315,13 +1510,15 @@ def render_gemm_arguments(
...
         Args:
+            instance_type (str): GEMM instance type.
             argument_template (str): Template for the GEMM operation arguments.
             epilogue_template (str): Template for the epilogue arguments.
             should_swap_xw (bool): Determines whether X, W operands should be swapped. If True, applies an explicit
             transpose operation to X and W.
             X (IRNode): The X input tensor.
             W (IRNode): The W input tensor.
             Bias (IRNode): The bias tensor.
+            Meta (IRNode): The meta tensor.
```
Comment (kadeng) [inline torch/_inductor/codegen/cuda/gemm_template.py:1521]: What is a meta tensor in this context? Is it an optional argument or mandatory even for non-sparse GEMMs?

Comment (alexsamardzic) [reply torch/_inductor/codegen/cuda/gemm_template.py:1521]: It's again about sparse semi-structured MM: in this case, "meta" tensor encodes which two elements, out of each four successive elements of the original matrix, are actually provided through matrix A.

**Review by kadeng — APPROVED** [review]
lgtm!

**Comment by kadeng** [comment]
@alexsamardzic Looks like there's a land-time merge conflict. You might want to rebase. CI will have to run again.-

**Comment by alexsamardzic** [comment]
> @alexsamardzic Looks like there's a land-time merge conflict. You might want to rebase. CI will have to run again.-

Rebased, CI running now.

**Pull request closed without merging**